embeddings/hybrid-search.ipynb

```diff
@@ -0,0 +1,1124 @@
+    "\n",
+    "```\n",
+    "TF-IDF(t, d) = TF(t, d) * IDF(t)\n",
+    "```\n",
+    "\n",
+    "The TF-IDF vectorizer tries to put higher weight to signature words in the dataset, such as \"Shirts\" or \"Dino\", compared to trivial words, such as \"The\", \"a\" or \"of\", and counts how many times those signature words are used in the specified document. Each value of a sparse embedding represents a frequency of each word based on the counts.\n",
+    "\n",
+    "##What is Reciprocal Rank Fusion?\n",
+    "For merging the token-based and semantic search results, hybrid search uses [Reciprocal Rank Fusion (RRF)](https://plg.uwaterloo.ca/%7Egvcormac/cormacksigir09-rrf.pdf). RRF is an algorithm for combining multiple ranked lists of items into a single, unified ranking. It's a popular technique for merging search results from different sources or retrieval methods, especially in hybrid search systems and large language models.\n",
+    "\n",
+    "In case of the hybrid search of Vector Search, the dense distance and sparse distance are measured in different spaces and can't be directly compared to each other. Thus, RRF works effectively for merging and ranking the results from the two different spaces.\n",
+    "\n",
+    "Here's how RRF Works:\n",
+    "\n",
+    "- **Reciprocal rank**: For each item in a ranked list, calculate its reciprocal rank. This means taking the inverse of the item's position (rank) in the list. For example, the item ranked number one gets a reciprocal rank of 1/1 = 1, and the item ranked number two gets 1/2 = 0.5.\n",
+    "- **Sum reciprocal ranks**: Sum the reciprocal ranks for each item across all the ranked lists. This gives a final score for each item.\n",
+    "- **Sort by final score**: Sort the items by their final score in descending order. The items with the highest scores are considered the most relevant or important.\n",
+    "\n",
+    "In short, the items with higher ranks in both dense and sparse results will be pulled up to the top of the list. Thus, the item \"Google Blue Kids Sunglasses\" is at the top as it has higher ranks in both dense and sparse search results. Items like \"Google White Classic Youth Tee\" are low ranked as they only have ranks in the dense search result.\n",
+    "\n",
+    "## How the alpha parameter behaves\n",
+    "The example of how to use hybrid search sets the parameter `rrf_ranking_alpha` as 0.5 when creating the `HybridQuery` object. You can specify a weight on ranking the dense and sparse search results using the following values for `rrf_ranking_alpha`:\n",
+    "\n",
+    "- 1, or not specified: Hybrid search uses only dense search results and ignores sparse search results\n",
+    "- 0: Hybrid search uses only sparse search results and ignores dense search results\n",
+    "- 0 to 1: Hybrid search merges both results from dense and sparse with the weight specified by the value. 0.5 means they will be merged with the same weight\n"
+   ]
+  }
+ ],
```

Comment:
Suggest adding sections on:

- Conclusion (with links and resources for users to take away)
- Cleanup (to delete search indices, GCS buckets, and other created resources)

Reply:
Added!